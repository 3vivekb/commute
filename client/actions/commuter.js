import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addCommuter = createAction('add commuter')
export const createCommuter = (commuter, groupId) => {
  commuter.groupId = groupId
  commuter.id = uuid.v4()
  return [
    addCommuter(commuter),
    push(`/group/${groupId}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete commuter')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteCommuter = (commuterId, groupId) => [
  deleteLocally(commuterId),
  push(`/group/${groupId}`)
]

const updateLocally = createAction('update commuter')
/* const updateOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'update'
    }
  }) */ // TODO update on server
export const updateCommuter = (commuter, groupId) => [
  updateLocally(commuter),
  push(`/group/${groupId}`)
]
