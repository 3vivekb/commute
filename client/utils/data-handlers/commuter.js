
export function loadDataIfNeeded (props, component) {
  const {commuter, loadCommuter, loadSite, params} = props

  // check to see if data needs to be loaded
  // for the create commuter page, the params will be empty, but the site needs to load
  // for the edit commuter page, we also need to load the commuter
  if (
    (!params || !params.commuterId || (params.commuterId && !commuter)) &&
    !component.loadingInitialData
  ) {
    // data needs to be loaded
    component.loadingInitialData = true
    loadSite({id: params.siteId})
    if (params.commuterId && !commuter) {
      // on edit commuter page and commuter hasn't loaded
      loadCommuter({id: params.commuterId})
    } else {
      // loading of the site on create page will not trigger a component update
      // therefore, say data has loaded
      component.setState({dataLoading: false})
    }
    return
  }
  component.loadingInitialData = false
  component.setState({dataLoading: false})
}
