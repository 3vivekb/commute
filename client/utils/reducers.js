import update from 'react-addons-update'

import {addToEntityMap, addEntitiesToEntityMap, deleteFromEntityMap, deleteManyFromEntityMap} from './entities'

function addChildEntityToParent (parentMap, newEntity, parentIdField, childPluralName) {
  return update(parentMap, {
    [newEntity[parentIdField]]: {
      [childPluralName]: {
        $push: [newEntity._id]
      }
    }
  })
}

/**
 * Make handlers for children creation and deletion events
 *
 * @param  {Object} cfg  An configuration object structured as follows:
 *   - childPluralName (String): Plural name of child (parent's child id array)
 *   - childSingularName (String): Singular name of child (for making reducer title)
 *   - parentIdField (String): Id field of parent in child entity
 * @return {Object}     Reducers
 */
export function makeChildrenHandlers (cfg) {
  const reducers = {}
  reducers[`add ${cfg.childSingularName}`] = function (state, action) {
    const newEntity = action.payload
    return addChildEntityToParent(
      state,
      newEntity,
      cfg.parentIdField,
      cfg.childPluralName
    )
  }

  reducers[`add many ${cfg.childPluralName}`] = function (state, action) {
    const newEntities = action.payload
    let updatedState = state
    newEntities.forEach(newEntity => {
      updatedState = addChildEntityToParent(
        updatedState,
        newEntity,
        cfg.parentIdField,
        cfg.childPluralName
      )
    })
    return updatedState
  }

  reducers[`delete ${cfg.childSingularName}`] = function (state, action) {
    const entity = action.payload
    const childIds = state[entity[cfg.parentIdField]][cfg.childPluralName]
    return update(state, {
      [entity[cfg.parentIdField]]: {
        [cfg.childPluralName]: {
          $set: childIds.filter((id) => id !== entity._id)
        }
      }
    })
  }

  return reducers
}

/**
 * Make generic reducers for an entity type
 *
 * @param  {Object} cfg  An object with name and handlers parameters structured as follows:
 *   - name: must have singular and plural keys with the corresponding strings
 *   - handlers: An array of strings of handlers.  Possible values: ['add', 'add many', 'delete', 'set', 'set many']
 * @return {Object}      The generated reducers
 */
export function makeGenericReducerHandlers (cfg) {
  const reducers = {
    'log out' (state, action) {
      return {}
    }
  }
  if (cfg.handlers.indexOf('add') !== -1) {
    reducers[`add ${cfg.name.singular}`] = function (state, action) {
      return addToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('add many') !== -1) {
    reducers[`add many ${cfg.name.plural}`] = function (state, action) {
      return addEntitiesToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('delete') !== -1) {
    reducers[`delete ${cfg.name.singular}`] = function (state, action) {
      return deleteFromEntityMap(state, action.payload._id)
    }
  }

  if (cfg.handlers.indexOf('delete many') !== -1) {
    reducers[`delete many ${cfg.name.plural}`] = function (state, action) {
      return deleteManyFromEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('set') !== -1) {
    reducers[`set ${cfg.name.singular}`] = function (state, action) {
      return addToEntityMap(state, action.payload)
    }
  }

  if (cfg.handlers.indexOf('set many') !== -1) {
    reducers[`set ${cfg.name.plural}`] = function (state, action) {
      return addEntitiesToEntityMap(state, action.payload)
    }
  }

  return reducers
}
