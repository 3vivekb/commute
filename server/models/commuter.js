const Schema = require('mongoose').Schema

const geocodingPlugin = require('./plugins/geocode')
const trashPlugin = require('./plugins/trash')
const userPlugin = require('./plugins/user')

const modeType = {
  travelTime: Number
}

const schema = new Schema({
  modeStats: {
    bicycle: modeType,
    car: modeType,
    transit: modeType,
    walk: modeType,
    type: Schema.Types.Mixed
  },
  name: {
    required: true,
    type: String
  },
  siteId: {
    ref: 'Site',
    required: true,
    type: Schema.Types.ObjectId
  }
})

function postGeocodeHook (commuter) {
  console.log('commuter added or location changed, initiating polygon calculation')

  // import here to resolve circular import
  const models = require('./')
  const isochroneUtils = require('../utils/isochrones')

  models.Site.findOne({ _id: commuter.siteId, trashed: undefined, user: commuter.user })
    .exec()
    .then((site) => {
      if (site.calculationStatus === 'successfully') {
        models.Polygon.find({ _id: commuter.siteId })
          .exec()
          .then((polygons) => {
            const siteIsochrones = {}
            polygons.forEach((polygon) => {
              if (!siteIsochrones[polygon.mode]) {
                siteIsochrones[polygon.mode] = {
                  features: []
                }
              }
              siteIsochrones[polygon.mode].features.push(polygon)
            })

            if (site.calculationStatus === 'successfully') {
              isochroneUtils.calculateIsochroneStatsForCommuter(commuter, siteIsochrones)

              console.log('commuter stats calculated')

              commuter.save()
            }
          })
          .catch((err) => {
            console.error('error calculating commuter stats:', err)
          })
      }
    })
    .catch((err) => {
      console.error('error calculating commuter stats:', err)
    })
}

schema.plugin(geocodingPlugin(postGeocodeHook))
schema.plugin(trashPlugin)
schema.plugin(userPlugin)

module.exports = schema
