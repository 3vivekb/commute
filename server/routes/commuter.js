const models = require('../models')
const endpointFactory = require('../utils/restEndpoints')

module.exports = function makeRoutes (app, jwt) {
  const baseCfg = {
    model: models.Commuter,
    name: 'commuter'
  }
  endpointFactory.makeRestEndpoints(
    app,
    jwt,
    Object.assign(
      {
        commands: {
          'Collection GET': {},
          'Collection POST': {},
          GET: {},
          DELETE: {},
          PUT: {}
        }
      },
      baseCfg
    )
  )
  endpointFactory.makePublicRestEndpoints(app, baseCfg)
}
