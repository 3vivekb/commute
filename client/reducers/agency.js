import {makeGenericReducerHandlers} from '../utils/reducers'

export const reducers = makeGenericReducerHandlers({
  handlers: ['add', 'delete', 'set', 'set all'],
  name: {
    singular: 'agency',
    plural: 'agencies'
  }
})

export const initialState = {}
