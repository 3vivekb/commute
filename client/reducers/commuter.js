import {addToEntityMap} from '../utils/entities'
import {
  getGenericReducerInitialState,
  makeGenericReducerHandlers,
  updateTime
} from '../utils/reducers'

export const reducers = makeGenericReducerHandlers({
  handlers: ['add', 'add many', 'delete', 'set', 'set many'],
  name: {
    singular: 'commuter',
    plural: 'commuters'
  }
})

reducers['add group'] = function (state, action) {
  const newGroup = {...action.payload}
  let updatedState = {...state}
  newGroup.commuters.map((commuter) => {
    updatedState = addToEntityMap(updatedState, commuter)
  })
  return updateTime(updatedState)
}

reducers['append commuters'] = function (state, action) {
  let updatedState = state
  action.payload.commuters.forEach((commuter) => {
    updatedState = addToEntityMap(updatedState, commuter)
  })
  return updateTime(updatedState)
}

export const initialState = getGenericReducerInitialState()
