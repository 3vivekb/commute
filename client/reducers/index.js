import {handleActions} from 'redux-actions'

import * as agency from './agency'
import * as analysis from './analysis'
import * as commuter from './commuter'
import * as group from './group'
import * as organization from './organization'
import * as site from './site'
import * as user from './user'

export default {
  agency: handleActions(agency.reducers, agency.initialState),
  analysis: handleActions(analysis.reducers, analysis.initialState),
  commuter: handleActions(commuter.reducers, commuter.initialState),
  group: handleActions(group.reducers, group.initialState),
  organization: handleActions(organization.reducers, organization.initialState),
  site: handleActions(site.reducers, site.initialState),
  user: handleActions(user.reducers, user.initialState)
}
