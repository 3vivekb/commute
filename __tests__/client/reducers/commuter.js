/* globals describe, expect, it */

import {handleActions} from 'redux-actions'

import {commuterSal, mockCommuter, mockGroupCreation, mockStores} from '../../test-utils/mock-data'
import {makeGenericReducerTestCases} from '../../test-utils/reducers'

import * as commuter from '../../../client/reducers/commuter'

describe('client > reducers > commuter', () => {
  makeGenericReducerTestCases({
    handlers: {
      add: {
        initialState: commuter.initialState,
        payload: mockCommuter
      },
      delete: {
        initialState: mockStores.withAnalysisRun.commuter,
        payload: 'commuter-2'
      },
      set: {
        initialState: mockStores.withAnalysisRun.commuter,
        payload: commuterSal
      },
      'set all': {
        initialState: commuter.initialState,
        payload: [mockCommuter]
      }
    },
    initialState: commuter.initialState,
    name: {
      plural: 'commuters',
      singular: 'commuter'
    },
    reducers: commuter.reducers
  })

  it('should handle add group', () => {
    const reducer = handleActions(commuter.reducers, commuter.initialState)
    const newGroup = {...mockGroupCreation}
    const action = { payload: newGroup, type: 'add group' }
    const result = reducer(mockStores.withBlankOrganization.commuter, action)
    expect(result['new-commuter'].name).toEqual('Fake Commuter')
    expect(result).toMatchSnapshot()
  })

  it('should handle append commuters', () => {
    const reducer = handleActions(commuter.reducers, commuter.initialState)
    const action = {
      type: 'append commuters',
      payload: {
        groupId: 'group-2',
        commuters: [commuterSal]
      }
    }
    const result = reducer(mockStores.withAnalysisRun.commuter, action)
    expect(result['commuter-3']).toEqual(commuterSal)
    expect(result).toMatchSnapshot()
  })
})
