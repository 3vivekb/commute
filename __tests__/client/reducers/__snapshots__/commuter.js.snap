exports[`client > reducers > commuter generic reducer tests should handle add commuter 1`] = `
Object {
  "commuter-2": Object {
    "_id": "commuter-2",
    "address": "4321 XYZ Boulevard",
    "email": "luke@warm.cold",
    "groupId": "group-2",
    "lat": 38.916089,
    "lng": -76.970221,
    "name": "Luke Warm",
  },
}
`;

exports[`client > reducers > commuter generic reducer tests should handle delete commuter 1`] = `Object {}`;

exports[`client > reducers > commuter generic reducer tests should handle set commuter 1`] = `
Object {
  "commuter-2": Object {
    "_id": "commuter-2",
    "address": "4321 XYZ Boulevard",
    "email": "luke@warm.cold",
    "groupId": "group-2",
    "lat": 38.916089,
    "lng": -76.970221,
    "name": "Luke Warm",
  },
  "commuter-3": Object {
    "_id": "commuter-3",
    "address": "9876 ABC Ct",
    "email": "sal@a.mander",
    "lat": 38.915,
    "lng": -76.971,
    "name": "Sal A. Mander",
  },
}
`;

exports[`client > reducers > commuter generic reducer tests should handle set commuters 1`] = `
Object {
  "commuter-2": Object {
    "_id": "commuter-2",
    "address": "4321 XYZ Boulevard",
    "email": "luke@warm.cold",
    "groupId": "group-2",
    "lat": 38.916089,
    "lng": -76.970221,
    "name": "Luke Warm",
  },
}
`;

exports[`client > reducers > commuter should handle add group 1`] = `
Object {
  "new-commuter": Object {
    "_id": "new-commuter",
    "address": "4321 XYZ Boulevard",
    "email": "luke@warm.cold",
    "groupId": "new-group",
    "name": "Fake Commuter",
  },
}
`;

exports[`client > reducers > commuter should handle append commuters 1`] = `
Object {
  "commuter-2": Object {
    "_id": "commuter-2",
    "address": "4321 XYZ Boulevard",
    "email": "luke@warm.cold",
    "groupId": "group-2",
    "lat": 38.916089,
    "lng": -76.970221,
    "name": "Luke Warm",
  },
  "commuter-3": Object {
    "_id": "commuter-3",
    "address": "9876 ABC Ct",
    "email": "sal@a.mander",
    "lat": 38.915,
    "lng": -76.971,
    "name": "Sal A. Mander",
  },
}
`;
